torch/nn/functional.py

```diff
@@ -3991,6 +3992,15 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
         assert align_corners is not None
         if antialias:
             return torch._C._nn._upsample_bilinear2d_aa(input, output_size, align_corners, scale_factors)
+        # Two levels are necessary to prevent TorchScript from touching
+        # are_deterministic_algorithms_enabled.
+        if not torch.jit.is_scripting():
+            if torch.are_deterministic_algorithms_enabled() and input.is_cuda:
+                # Use slow decomp whose backward will be in terms of index_put
+                # importlib is required because the import cannot be top level
+                # (cycle) and cannot be nested (TS doesn't support)
+                return importlib.import_module('torch._decomp.decompositions').upsample_bilinear2d_vec(
+                    input, output_size, align_corners, scale_factors)
```

Comment:
is `import_lib` still necessary, if you're running it from inside of a region of `if not torch.jit.is_scripting`?

I would have thought that torchscript would know to not try to scriptify the region in this block, but maybe that's wrong

Reply:
Yes:

```
  File "/data/users/ezyang/d/pytorch/torch/jit/frontend.py", line 141, in <listcomp>
    stmts = [build_stmt(ctx, s) for s in stmts]
  File "/data/users/ezyang/d/pytorch/torch/jit/frontend.py", line 320, in __call__
    return method(ctx, node)
  File "/data/users/ezyang/d/pytorch/torch/jit/frontend.py", line 644, in build_If
    build_stmts(ctx, stmt.body),
  File "/data/users/ezyang/d/pytorch/torch/jit/frontend.py", line 141, in build_stmts
    stmts = [build_stmt(ctx, s) for s in stmts]
  File "/data/users/ezyang/d/pytorch/torch/jit/frontend.py", line 141, in <listcomp>
    stmts = [build_stmt(ctx, s) for s in stmts]
  File "/data/users/ezyang/d/pytorch/torch/jit/frontend.py", line 319, in __call__
    raise UnsupportedNodeError(ctx, node)
torch.jit.frontend.UnsupportedNodeError: import statements aren't supported:
  File "/data/users/ezyang/d/pytorch/torch/nn/functional.py", line 4002
                # importlib is required because the import cannot be top level
                # (cycle) and cannot be nested (TS doesn't support)
                import torch._decomp.decompositions as decomp
                ~~~~~~ <--- HERE
                return decomp.upsample_bilinear2d_vec(
                    input, output_size, align_corners, scale_factors)


----------------------------------------------------------------------

```

I think what is going on is that TorchScript is parsing the entire function body before eliminating is_scripting, so you can't use language features that the parser doesn't support.